File: frontend/lib/schemas.ts
import { z } from "zod";

export const ACCOUNT_PURPOSES = ["income", "spending", "savings", "other"] as const;

export const AccountSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  purpose: z.enum(ACCOUNT_PURPOSES),
  is_tracked: z.boolean(),
});
export type Account = z.infer<typeof AccountSchema>;

export const SkippedRowSchema = z.object({
  row_number: z.number().int(),
  reason: z.string(),
});
export type SkippedRow = z.infer<typeof SkippedRowSchema>;

export const StatementUploadResponseSchema = z.object({
  statement_id: z.string().uuid(),
  status: z.string(),
  transactions_imported: z.number().int(),
  transactions_skipped: z.number().int(),
  skipped_rows: z.array(SkippedRowSchema),
});
export type StatementUploadResponse = z.infer<typeof StatementUploadResponseSchema>;

export const CATEGORY_TYPES = ["income", "fixed_expense", "variable_expense"] as const;

export const CategorySchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  type: z.enum(CATEGORY_TYPES),
});
export type Category = z.infer<typeof CategorySchema>;

export const TransactionSchema = z.object({
  id: z.string().uuid(),
  txn_date: z.string(),
  raw_description: z.string(),
  amount: z.string(), // Decimal, serialized as a string -- never parse as float
  direction: z.enum(["debit", "credit"]),
  account_name: z.string(),
  category_id: z.string().uuid().nullable(),
  category_name: z.string().nullable(),
  confidence: z.string().nullable(),
  needs_review: z.boolean(),
  is_internal_transfer: z.boolean(),
});
export type Transaction = z.infer<typeof TransactionSchema>;

export const CategoryTotalSchema = z.object({
  category_id: z.string().uuid(),
  name: z.string(),
  type: z.enum(CATEGORY_TYPES),
  total: z.string(),
});
export type CategoryTotal = z.infer<typeof CategoryTotalSchema>;

export const MonthlySummarySchema = z.object({
  month: z.string(),
  total_income: z.string(),
  total_fixed_expenses: z.string(),
  total_variable_expenses: z.string(),
  discretionary_income: z.string(),
  pending_review_count: z.number().int(),
  by_category: z.array(CategoryTotalSchema),
});
export type MonthlySummary = z.infer<typeof MonthlySummarySchema>;

// Client-side pre-validation, so an obviously bad file never leaves the browser.
export const ALLOWED_EXTENSIONS = ["csv", "xlsx", "xls", "pdf"] as const;
export const MAX_FILE_SIZE_BYTES = 10 * 1024 * 1024; // 10 MB

export function validateStatementFile(file: File): string | null {
  if (file.size === 0) return "the file is empty";
  if (file.size > MAX_FILE_SIZE_BYTES) return "file must be under 10 MB";
  const ext = file.name.split(".").pop()?.toLowerCase();
  if (!ext || !(ALLOWED_EXTENSIONS as readonly string[]).includes(ext)) {
    return "unsupported file type -- use CSV, XLSX, or PDF";
  }
  return null;
}

File: app/schemas.py
"""Pydantic request/response models for the API layer."""

from __future__ import annotations

import enum
import uuid
from datetime import date
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_serializer, field_validator


class MoneyModel(BaseModel):
    """Base for any response model carrying Decimal money fields.

    Pydantic's default JSON encoding can convert Decimal to float, which
    silently loses precision -- never acceptable for currency. This
    forces every Decimal field to serialize as a string instead.
    """

    @field_serializer("*", when_used="json")
    def _decimal_as_str(self, value: object) -> object:
        return str(value) if isinstance(value, Decimal) else value


class AccountOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    purpose: str
    is_tracked: bool

    @field_validator("purpose", mode="before")
    @classmethod
    def _coerce_enum(cls, v: object) -> object:
        return v.value if isinstance(v, enum.Enum) else v


class AccountCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    purpose: Literal["income", "spending", "savings", "other"]
    is_tracked: bool = True


class AccountUpdateRequest(BaseModel):
    is_tracked: bool


class CategoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    type: str

    @field_validator("type", mode="before")
    @classmethod
    def _coerce_enum(cls, v: object) -> object:
        return v.value if isinstance(v, enum.Enum) else v


class CategoryCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    type: Literal["income", "fixed_expense", "variable_expense"]


class SkippedRowOut(BaseModel):
    row_number: int
    reason: str


class StatementUploadResponse(BaseModel):
    statement_id: uuid.UUID
    status: str
    transactions_imported: int
    transactions_skipped: int
    skipped_rows: list[SkippedRowOut] = Field(default_factory=list)


class TransactionOut(MoneyModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    txn_date: date
    raw_description: str
    amount: Decimal
    direction: str
    account_name: str
    category_id: uuid.UUID | None
    category_name: str | None
    confidence: Decimal | None
    needs_review: bool
    is_internal_transfer: bool

    @field_validator("direction", mode="before")
    @classmethod
    def _coerce_enum(cls, v: object) -> object:
        # Defensive: works whether the ORM gives us the enum member or a raw string.
        return v.value if isinstance(v, enum.Enum) else v


class CategoryCorrectionRequest(BaseModel):
    category_id: uuid.UUID


class CategoryTotalOut(MoneyModel):
    model_config = ConfigDict(from_attributes=True)

    category_id: uuid.UUID
    name: str
    type: str
    total: Decimal

    @field_validator("type", mode="before")
    @classmethod
    def _coerce_enum(cls, v: object) -> object:
        return v.value if isinstance(v, enum.Enum) else v


class MonthlySummaryOut(MoneyModel):
    month: date
    total_income: Decimal
    total_fixed_expenses: Decimal
    total_variable_expenses: Decimal
    discretionary_income: Decimal
    pending_review_count: int
    by_category: list[CategoryTotalOut]
